uplink/ecclient/client.go

```diff
@@ -473,30 +474,81 @@ func (lr *lazyPieceRanger) Size() int64 {
 // Range implements Ranger.Range to be lazily connected
 func (lr *lazyPieceRanger) Range(ctx context.Context, offset, length int64) (_ io.ReadCloser, err error) {
 	defer mon.Task()(&ctx)(&err)
-	ps, err := lr.dialPiecestore(ctx, &pb.Node{
+
+	reader := &lazyPieceReader{}
+	reader.lazyPieceRanger = lr
+	reader.ctx = ctx
+	reader.offset = offset
+	reader.length = length
+	return reader, nil
+}
+
+type lazyPieceReader struct {
+	*lazyPieceRanger
+	ctx    context.Context
+	offset int64
+	length int64
+
+	mu sync.RWMutex
+
+	dialed bool
+	closed bool
+	piecestore.Downloader
+	client *piecestore.Client
+}
+
+func (lr *lazyPieceReader) Read(data []byte) (_ int, err error) {
+	lr.mu.RLock()
+	defer lr.mu.RUnlock()
+
+	if lr.closed {
+		return 0, io.EOF
+	}
+	if !lr.dialed {
+		err = lr.dial()
+		if err != nil {
+			return 0, err
+		}
+	}
+	if lr.dialed && lr.Downloader == nil {
+		return 0, io.EOF
+	}
+
+	return lr.Downloader.Read(data)
+}
+
+func (lr *lazyPieceReader) dial() (err error) {
+	defer mon.Task()(&lr.ctx)(&err)
+	lr.dialed = true
+	ps, err := lr.dialPiecestore(lr.ctx, &pb.Node{
 		Id:      lr.limit.GetLimit().StorageNodeId,
 		Address: lr.limit.GetStorageNodeAddress(),
 	})
 	if err != nil {
-		return nil, err
+		return err
 	}
+	lr.client = ps
 
-	download, err := ps.Download(ctx, lr.limit.GetLimit(), lr.privateKey, offset, length)
+	download, err := ps.Download(lr.ctx, lr.limit.GetLimit(), lr.privateKey, lr.offset, lr.length)
 	if err != nil {
-		return nil, errs.Combine(err, ps.Close())
+		return errs.Combine(err, ps.Close())
 	}
-	return &clientCloser{download, ps}, nil
+	lr.Downloader = download
+	return nil
 }
 
-type clientCloser struct {
-	piecestore.Downloader
-	client *piecestore.Client
-}
+func (lr *lazyPieceReader) Close() error {
+	lr.mu.Lock()
+	defer lr.mu.Unlock()
```

Comment:
Is the purpose of mutex to handle a concurrent call to `Close` with `Read`? It seems like that's its benefit. To me, a simple comment at the top of the struct stating that "lazyPieceReader does not support concurrent calls to `Read` or `Close`. Only make sequential calls to Read, and only call Close after all calls to Read are finished."

That seems like would make it possible to get rid of the mutex, but I'm ok either way.

Reply:
Not sure why, because it seems like we close everything at the very end of reading from a download, but I was running into a data race on reading and writing to `lazyPieceReader.closed`, so it seemed there was some concurrent reads and closes?